### PR TITLE
fix missing events after transformation

### PR DESCRIPTION
**Pull request by MGathier**
when the last transaction in a segment contained multiple events and the last event was not updated in the transaction the last events were not copied into the new version of the segment. Also, when a domain event was deleted, the event was no longer present in the index as the replacement event had no aggregate type.

### CODE REVIEW

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java [hunk 1]
```diff
@@ -126,55 +129,82 @@ private void closeTransactionIterator() {
...
+                        .setAggregateIdentifier(originalEvent.getAggregateIdentifier())
+                        .setAggregateSequenceNumber(originalEvent.getAggregateSequenceNumber())
+                        .setAggregateType(originalEvent.getAggregateType().isEmpty() ? "" : "-")
+                        .setPayload(EMPTY_PAYLOAD)
+                        .build();
+        }
+        return replacementEvent;
+    }
+
+    private SerializedTransactionWithToken originalTransaction() {
```
Comment (m1l4n54v1c) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java:181]: This has a signature that makes you think it's an accessor method. Instead, it changes the state.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java [hunk 2]
```diff
@@ -126,55 +129,82 @@ private void closeTransactionIterator() {
...
+    }
+
+    private void write(List<Event> events) throws IOException {
+        SegmentWriter segmentWriter = segmentWriterRef.get();
+        segmentWriter.write(events);
+    }
+
+    private void writeRemainingEvents() throws IOException {
+        // first complete the remaining events in the original transaction
+        SerializedTransactionWithToken originalTX = originalTransactionRef.get();
```
Comment (m1l4n54v1c) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java:196]: Why don't we call here `originalTransaction()`?

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java [hunk 3]
```diff
@@ -126,55 +129,82 @@ private void closeTransactionIterator() {
...
+        if (originalTX != null) {
+            while (transformedTransaction.size() < originalTX.getEventsCount()) {
+                transformedTransaction.add(originalTX.getEvent(transformedTransaction.size()));
```
Comment (m1l4n54v1c) [inline axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java:199]: This will always add the last event?

**Review by m1l4n54v1c — CHANGES_REQUESTED** [review]

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java [hunk 4]
```diff
@@ -126,55 +129,82 @@ private void closeTransactionIterator() {
...
         }).subscribeOn(Schedulers.boundedElastic());
     }
+
+    private Event transform(Event originalEvent, Event replacementEvent) {
+        if (Event.getDefaultInstance().equals(replacementEvent)) {
+            return Event.newBuilder()
+                        .setTimestamp(originalEvent.getTimestamp())
+                        .setAggregateIdentifier(originalEvent.getAggregateIdentifier())
+                        .setAggregateSequenceNumber(originalEvent.getAggregateSequenceNumber())
+                        .setAggregateType(originalEvent.getAggregateType().isEmpty() ? "" : "-")
```
[the remark this answers is not in the document]
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java:174]: Changed to keep the original aggregate type. If aggregate type is empty if will not be included in the index.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java (hunk 1 above)
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java:181]: renamed the operation.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java (hunk 2 above)
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java:196]: in this case, I only want to check if there is a partly transformed transaction. Don't want to move to another transaction if there is none.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java (hunk 3 above)
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java:199]: Changed the loop so the logic is easier to read.

File: axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (m1l4n54v1c) [reply axonserver/src/main/java/io/axoniq/axonserver/localstorage/file/DefaultSegmentTransformer.java]: But if the original aggregate type is empty, it shouldn't be included in the index in the first place?

**Review by m1l4n54v1c — APPROVED** [review]
My concerns are addressed.

**Pull request merged**